### ISSUE TITLE
Adding `-pthread` to LIBS leads to issues

### ISSUE THREAD

**Issue by Flamefire** [easybuilders/easybuild-framework#3154]
Power9 strikes again: Using some functions from pthread (e.g. `pthread_atfork`) requires linking of `lpthread`. E.g. https://github.com/scipy/scipy/issues/11323 came up.

I traced the problem to EasyBuild setting `LIBS=-lpthread`, see https://github.com/scipy/scipy/issues/11323#issuecomment-571697982

In short: Pythons distutils compiles extensions with the same flags as determined when configuring Python. Python is configured using autotools which uses `LIBS`, distutils does not seem to do so. Now Pythons configure thinks, it does not need any flag for pthread support and hence does not record any for distutils. When an extension now uses pthreads it won't link against libpthread directly which leads to the above failure on Power9.

What was the reasoning on unconditionally adding `-lm` and `-pthread` to `LIBS`? Isn't that what configure and the likes is for?

**Comment by ocaisa** [comment]
I did a bit of digging and the relevant LoC start at https://github.com/easybuilders/easybuild-framework/blob/develop/easybuild/tools/toolchain/compiler.py#L225

These lines are 7 years old, so I expect that we are looking at unintended consequences. Given that `distutils` uses the Python compile time flags (but not `LIBS`) is the fix here something that should go into the Python easyblock rather than in the framework?

**Comment by Flamefire** [comment]
Maybe: What about not adding it to LIBS for Python? Would that be possible? Or check/assume availability of `-pthread` and add that the CFLAGS? Raise an issue in CPython that LIBS is ignored?

**Comment by ocaisa** [comment]
Is `-pthread` portable among "all" compilers? If so, we could either just add it to CFLAGS in the Python easyblock (but wouldn't you still have a link time problem for the packages?) or add it to `LDSHARED` in the Python package easyblock and always set it? I'm not entirely sure which is the best option though?

**Comment by Flamefire** [comment]
To "most" compilers as they converged to that. But no this is not a solution  as a) it is needed for compiling and linking and b) you'll run into the same issue: https://bugs.python.org/issue31769

If at all you'd need to somehow make the distutils use the parameter when linking extensions in the PythonExtension EB

**Comment by ocaisa** [comment]
Ok, then as far as I can tell the solution for linking here is to manipulate `LDSHARED` to include the flag and always set that envvar for extensions (the groundwork for that is already in place), and also add it CFLAGS (and friends)? I guess we should check in the easyblock that it is valid before doing that.

**Comment by ocaisa** [comment]
Checking/setting of LD_SHARED is done by default for recent Python versions (see https://github.com/easybuilders/easybuild-easyblocks/blob/develop/easybuild/easyblocks/generic/pythonpackage.py#L495) so to fix the linking you could just add `LDSHARED="$LDSHARED -pthread"` as a `preconfigopts` for whatever package is not linking.

**Comment by Flamefire** [comment]
That code seems to have a bug. I checked the log for a python installation and found `Python's value for $LDSHARED ('gcc -pthread -shared -Wl,-z,relro') uses current $CC value ('gcc'), not touching it`. Then I executed `ml Python/3.7.4-GCCcore-8.3.0; python -c "from distutils import sysconfig; print(sysconfig.get_config_var('LDSHARED'))"` and got `gcc -shared <lot's of -L<folder>>` but no `-pthread`.

Checking https://github.com/easybuilders/easybuild-easyblocks/blob/126c726f351a6ef922b87f946c9e1d22dc6bbef7/easybuild/easyblocks/generic/pythonpackage.py#L509 I noticed that it uses the **hosts** sysconfig which is obviously wrong, isn't it? So this isn't doing what it is supposed to...

> manipulate LDSHARED to include the flag and always set that envvar for extensions

should work. Not the `-pthread` and `-lpthread` are different. The latter must be placed after the object files, while the former can be placed anywhere

> and also add it CFLAGS (and friends) when compiling extensions?

maybe required but usually (now always?) is not

> so to fix the linking you could just add LDSHARED="$LDSHARED -pthread" as a preconfigopts for whatever package is not linking.

I'm rather looking for a generic solution, e.g. by changing the EasyBlock.

**Comment by ocaisa** [comment]
>  That code seems to have a bug.

You're right, there is a bug there and it is also interesting to see the -pthread flag there from the OS.

> Note the -pthread and -lpthread are different. 

> maybe required but usually (now always?) is not
 
Yeah, I saw that. I think we can figure out a way to ensure the `-pthread` is accepted by `$CC`, then add it to `LDSHARED` in the pythonpackage easyblock.

@boegel Is there anything we are missing here?

**Comment by Flamefire** [comment]
> interesting to see the -pthread flag there from the OS.

Why is this unexpected? That is exactly what should be done and only isn't done by EB installed Pythons because LIBS=-lpthread is passed during configure. The system python is not build with that and hence correctly detects and adds the flag to its config

**Comment by ocaisa** [comment]
Ok, I was going to ask that earlier, so just popping out the `LIBS` envvar when configuring Python would also solve this problem?

**Comment by Flamefire** [comment]
I'm 95% sure yes. Therefore the issue name. We should even add a sanity check that LDSHARED contains pthread (or more generically: The threading lib or `-pthread`)

I'd rather say no. What are we doing with LDSHARED? Only replacing the compiler? At least that looks like it is what was intended but https://github.com/easybuilders/easybuild-easyblocks/blob/126c726f351a6ef922b87f946c9e1d22dc6bbef7/easybuild/easyblocks/generic/pythonpackage.py#L517 replaces everything O.o   
There is also a CXX equivalent in my Python 3.7.4 config LDCXXSHARED which probably would also need to be handled.

I don't even understand why the replacement should be done. Isn't that only for linking? IMO that was (mostly? completely?) compiler independent

**Comment by ocaisa** [comment]
> I don't even understand why the replacement should be done. Isn't that only for linking? IMO that was (mostly? completely?) compiler independent

It's there because sometimes Python is built with GCC(core) and the extension is compiled with the compiler of the toolchain (e.g. intel), when this is the case, you need to link with the same compiler family or you run into problems. It replaces everything because `LDSHARED` may have options for the original linker that are not valid (and kick an error).

**Comment by Flamefire** [comment]
Hmm... Wouldn't you want the Python package built with the same compiler due to e.g. Scipy and Numpy being less optimized otherwise?

Anyway I wouldn't add -pthread there. Maybe the compilers threading lib.

**Comment by ocaisa** [comment]
> Hmm... Wouldn't you want the Python package built with the same compiler due to e.g. Scipy and Numpy being less optimized otherwise?

Python is built with GCCcore, the Scipy package is built with Intel (and in particular MKL) exactly for performance.

> Anyway I wouldn't add -pthread there. Maybe the compilers threading lib.

The problem with adding the threading lib is that it has to come at the end of the link command, and we can't influence that with `LDSHARED`, `-pthread` is our only option in this case.

**Comment by Flamefire** [comment]
Ah, thanks, my mistake.

True. Hm... I guess the toolchain would need to announce -pthread support or we check the threading lib to match pthread and add -pthread if it does.

**Comment by boegel** [comment]
cc'ing @Micket since he was involved with the `LDSHARED` stuff in `PytonPackage`, it indeed seems that we should be using `pyton -c "from distutils.sysconfig import get_config_vars; print(get_config_vars('LDSHARED'))"` to determine the `$LDSHARED` of the active Python...

Also cc'ing @edmondac and @branfosj since they've been submitting various fixes related to POWER (including something related to `pthread`, but I forgot where exactly).

**Comment by Micket** [comment]
I thought we already were. Yes, we must absolutely pick up the this from the python dependency itself, not whatever EB uses.

Alternatively; We could just set LDSHARED to what we think it should be. I mean; that's what we end up doing for *every* pythonpackage compiled with intel anyway, so it's not just some rare exception. So I think that proves that those extra link flags that we "keep" for the foss toolchain are demonstrably not necessary. (this is what my original PR for LDSHARED did actually, but @boegel forced me to change it :smile:  )

**Comment by branfosj** [comment]
We've seen the `pthread` on POWER issue a couple of times - in an older version of PyTorch and in the cryptography Python library. Each time we've fixed it at the place we have seen it.

Getting a fix in to solve this would be good. I like the idea of setting and using LDSHARED.

**Comment by Flamefire** [comment]
Yeah. I'm just unsure if *replacing* LDSHARED is the right way to go. I feel it would be better to use full toolchain pythons and not set the threading lib in LIBS (or add it afterwards to the installed config)

What if there are some parameters in LDSHARED that are required or set to something site specific?

**Comment by Micket** [comment]
Undoing all the recent work to move Python to GCCcore is not an option I.M.H.O.
We must be able to expect any library from GCCcore to be compatible with iccifort, be it libpython or zlib. This what-if could be applied to any library.

Normally, think like pkg-config, python-config, wouldn't include which compiler was used, so LDSHARED is already to odd one out here. It basically always be equal to `$CC` + `python-config --ldflags` + `-shared` (or whatever is the compiler specific equivalent). Though, I'm not sure we need to add LDFLAGS manually, as I think they might be added anyway; which just leaves `$CC -shared` (which has at least been shown to work for all python packages we have in EB right now (with intel on x86_64 platforms)).

**Comment by Flamefire** [comment]
> This what-if could be applied to any library.

Sure. I still think Python is special here as distutils is a build system which ensures all dependencies are built like Python was and there can be other assumptions based on that. Maybe someone with more experience with Python package development can comment on whether this is true.

> Normally, think like pkg-config, python-config, wouldn't include which compiler was used

Problem is: How could you include flags, when you don't include the compiler? This is a big disadvantage of pkgconfig in that it allows basically adding any flags, but those may or may not be supported by the compiler that is then used to consume that library.

We can try and patch LDSHARED for PythonPackages and add `-pthread` for supported toolchains or add it to the python config (during or after installation) which seems to be more correct. Let's then see if other problems from mixing compilers come up.

**Comment by Micket** [comment]
> Sure. I still think Python is special here as distutils is a build system which ensures all dependencies are built like Python was and there can be other assumptions based on that. Maybe someone with more experience with Python package development can comment on whether this is true.

distutils is absolutely not strict enough to expect python packages not to be messing it up just as much as with any other build scripts. Have a look at GPAW for example.

I don't think build systems are allowed to ever be special. Python just can't be allowed to determine how all it's packages should be linked. It certainly can't do so for anything using SWIG, boost-python, pybind, etc.

Also, https://github.com/python/cpython/blob/master/Lib/distutils/sysconfig.py#L206 using the LDSHARED environment variable is how distutils expects you to customize the link command; it's the only reason the environment variable exists.
It's not much different from setting CC, CFLAGS, 

> Problem is: How could you include flags, when you don't include the compiler?

It's not like including the compiler command improves anything; It just means that only one compiler may ever be used, which is always unacceptable.

> We can try and patch LDSHARED for PythonPackages and add `-pthread` for supported toolchains or add it to the python config (during or after installation) which seems to be more correct. Let's then see if other problems from mixing compilers come up.

distutils will be adding LDFLAGS as well, so, -pthread should could possible better belong there.
https://github.com/python/cpython/blob/master/Lib/distutils/sysconfig.py#L212

The question that I'm still not sure about is
1. Do we append -pthread to the link flags for python packages?
2. Do we compile python itself with -pthread to start with so that it shows up for pythonpackage's by itself?

I accidentally also found this relevant PR: https://github.com/python/cpython/pull/14950

**Comment by Flamefire** [comment]
> It's not like including the compiler command improves anything; It just means that only one compiler may ever be used, which is always unacceptable.

It does: It is a clean-room approach ensuring that everything is compatible. vector-of-bool recently used a similar approach: A build system using only source distributions where everything will be build by the same compiler and flags.

It find that reasonable: New compilers introduce other defaults (e.g. C++ std level) and may render binaries incompatible.

**Comment by boegel** [comment]
@Flamefire What's the best way forward for this?

**Comment by Flamefire** [comment]
Multiple options and I don't think we converged to one. To summarize (hopefully):

1. Not adding `-pthread` to LIBS ever
1. Removing the threading lib from LIBS when building Python
1. Adding `-pthread` to LDSHARED when building a Python package ( and opt. to LDCXXSHARED)
1. Patch the python installation to include `-phread` in its config
1. Add `-pthread` to LDFLAGS env variable when building PythonPackages

And in all cases: Check the Intel toolchain if we break anything due to manipulation of LDSHARED (which is already happening)
